### PR TITLE
Crud ul

### PR DESCRIPTION
**Pull request by mnadiyaka**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -57,10 +57,9 @@
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <version>4.12</version>
-            <scope>test</scope>
```
Comment (tarasVarvariuk) [inline pom.xml:63]: Why did you remove JUnit dependency?

File: src/main/java/com/sigma/Application.java [hunk 2]
```diff
@@ -2,10 +2,18 @@
...
 @SpringBootApplication
 public class Application {
     public static void main(String[] args) {
         SpringApplication.run(Application.class, args);
     }
+
+//    @Bean
+//    public PasswordEncoder passwordEncoder() {
+//        return new BCryptPasswordEncoder(BCryptPasswordEncoder.BCryptVersion.$2B);
+//    }
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/Application.java]: Try not to collect unused code. If it's non needed any more - remove it. You'll find it in commit history later

File: src/main/java/com/sigma/JWTUtil.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.security.authentication.AuthenticationManager;
+
+import java.util.Base64;
+import java.util.Date;
+
+@Slf4j
+@RequiredArgsConstructor
+public class JWTUtil {
+
+    private final AuthenticationManager authenticationManager;
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java]: If it's unit class it's expected to have no external dependencies and private constructor. Why is this authenticationManager here?

File: src/main/java/com/sigma/JWTUtil.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    public static String generateJWT(User user) {
+        Algorithm algorithm = Algorithm.HMAC256("secret".getBytes()); //TODO: put in file
+        String access_token = JWT.create()
+                .withSubject(user.getId().toString())
+                .withClaim("role", user.getRole().toString())
+                .withIssuer("http://localhost:8080/")
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java]: Hosts, timestamps, ids, secrets should all be stored in external property file

File: src/main/java/com/sigma/JWTUtil.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public static void verify(User user, final String token) {
+
+        String[] chunks = token.split("\\.");
+
+        Base64.Decoder decoder = Base64.getUrlDecoder();
+
+        String header = new String(decoder.decode(chunks[0]));
+        String payload = new String(decoder.decode(chunks[1]));
+
+        log.info("user {}, password {}", user.getUsername(), user.getPassword());
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java]: You are missing key part here. You should decode token with secret to verify it. Then confirm expirationTs to confirm it hasn't expired.

File: src/main/java/com/sigma/configuration/SecurityConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder(BCryptPasswordEncoder.BCryptVersion.$2B);
+    }
+
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth.inMemoryAuthentication()
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/SecurityConfiguration.java]: I'm not sure what's going on here at all. What are you trying to achieve provide hardcoded credentials?

File: src/main/java/com/sigma/JWTUtil.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                .withIssuer("http://localhost:8080/")
+                .withIssuedAt(new Date(System.currentTimeMillis()))
+                .withExpiresAt(new Date(System.currentTimeMillis() + 5 * 60 * 1000)) //TODO: change time
+                .sign(algorithm);
+
+
+        return access_token;
+    }
+
+    public static void verify(User user, final String token) {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java]: There is a good convention that you should make every variable final if you are not going to modify it explicitly.

File: src/main/java/com/sigma/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    public SignUpUserResponseDto signUp(@RequestBody SignUpUserDto signUpDto) {
+        return userService.createUser(signUpDto);
+    }
+
+    @PostMapping("/login")
+    public SignInUserResponseDto loginUser(@RequestBody SignInUserDto user) throws AuthenticationException {
+        User myUser = userService.findUserByUsername(user.getUsername());
+        List<User> users = userService.getAllUsers();
+//        String test = myUser.getPassword();
+//        String test1 = passwordEncoder.encode(user.getPassword());
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Again remove unused code

File: src/main/java/com/sigma/controller/UserController.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @DeleteMapping("/users/{user_id}/delete")
+    public String deleteUser(@PathVariable("user_id") Long userId) {
+        userService.deleteUser(userId);
+        return "deleted user";
+    }
+
+    @GetMapping("/admin")
+    public String admin_in() {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Method name violates common Java conventions

File: src/main/java/com/sigma/filter/AuthFilter.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+
+
+@Slf4j
+@RequiredArgsConstructor
+public class AuthFilter extends UsernamePasswordAuthenticationFilter {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/filter/AuthFilter.java]: Not what is expected here. Should extend regular HttpFilter. No need for any AuthenticationManager or what so ever. Just get token from Headers and verify it.

File: src/main/java/com/sigma/service/impl/LocationServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    public Location findLocationById(Long locationId) {
+        log.info("Searching for location with id {}", locationId);
+        return locationRepository.findById(locationId).orElseThrow(() -> new EntityNotFoundException());
+    }
+
+    @Override
+    public Location createLocation(Location location) {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/service/impl/LocationServiceImpl.java]: We suggested to use DTO for CRUD

File: src/main/java/com/sigma/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public User findUserById(Long userId) {
+        log.info("Searching for user with id {}", userId);
+        return userRepository.findById(userId).orElseThrow(() -> new EntityNotFoundException());
+    }
+
+    @Override
+    public User findUserByUsername(String username) {
+        log.info("Searching for user with username {}", username);
+        if (userRepository.findByUsername(username) == null)
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/service/impl/UserServiceImpl.java]: There's a common convention to add curly brackets even when 'if' contains single line

File: src/main/java/com/sigma/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public SignUpUserResponseDto createUser(SignUpUserDto signUpDto) {
+        log.info("Creating new user {}", signUpDto.getUsername());
+        if (userRepository.findByUsername(signUpDto.getUsername()) != null)
+            return new SignUpUserResponseDto("failure", "user already exists");
+
+        User user = new User(signUpDto.getUsername(), passwordEncoder.encode(signUpDto.getPassword()), signUpDto.getRole());
+        userRepository.save(user);
+        return new SignUpUserResponseDto("success", "user created");
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/service/impl/UserServiceImpl.java]: Any statuses like 'success', 'user created' etc. Should be constants.

File: src/main/resources/application.properties [hunk 14]
```diff
@@ -7,6 +7,6 @@ spring.sql.init.mode=always
 spring.jpa.database=MYSQL
```
Comment (tarasVarvariuk) [inline src/main/resources/application.properties]: I think we suggested to use YAML format rather then .properties. Try using yaml config, it's more readable.

File: src/test/java/com/sigma/service/LocationServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import javax.transaction.Transactional;
+
+@SpringBootTest
+public class LocationServiceTest {
+
+    @Autowired
+    private LocationRepository locationRepository;
```
Comment (tarasVarvariuk) [inline src/test/java/com/sigma/service/LocationServiceTest.java:16]: Let's discuss unit tests after. For now suggested to get familiar with Mockito framework and concept of mocks in general.

File: target/maven-status/maven-compiler-plugin/compile/default-compile/createdFiles.lst [hunk 16]
```diff
@@ -1,13 +1,25 @@
-com\sigma\model\Category.class
-com\sigma\model\Role.class
 com\sigma\model\User.class
-com\sigma\model\Team.class
 com\sigma\Application.class
 com\sigma\repository\LocationRepository.class
-com\sigma\model\Participant.class
+com\sigma\service\impl\UserServiceImpl.class
```
Comment (tarasVarvariuk) [inline target/maven-status/maven-compiler-plugin/compile/default-compile/createdFiles.lst]: Why is target in pull request again? I should only see it removed.

File: pom.xml (hunk 1 above)
Comment (mnadiyaka) [reply pom.xml:63]: it works without it, with this depwndancy i had exception about different sourses (or something similar), i think in sprind framework it olready included

File: src/main/java/com/sigma/Application.java (hunk 2 above)
Comment (mnadiyaka) [reply src/main/java/com/sigma/Application.java]: ok

File: src/main/java/com/sigma/JWTUtil.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public static void verify(final String token, String secret, String issuer) {
+        Algorithm algorithm = Algorithm.HMAC256(secret.getBytes());
+        JWTVerifier verifier = JWT.require(algorithm)
+                .withIssuer("http://localhost:8080/")
+                .build();
+
+        DecodedJWT jwt = JWT.decode(token);
+        if( jwt.getExpiresAt().before(new Date())) {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java]: First verify token itself. Then 'expiresAt'

File: src/main/java/com/sigma/configuration/SecurityConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
+@EnableWebSecurity
+public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/SecurityConfiguration.java]: No need for Spring security config

File: src/main/java/com/sigma/controller/UserController.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+@RequiredArgsConstructor
+@RequestMapping("")
+@Slf4j
+public class UserController {
+
+    private final String SUCCESS = "success";
+    private final String FAILURE = "failure";
+
+    @Autowired
+    private UserService userService;
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Let's @Autowire everything through constructor

File: src/main/java/com/sigma/controller/UserController.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @PostMapping("/signUp")
+    public SignUpUserResponseDto signUp(@RequestBody SignUpUserDto signUpDto) {
+        return userService.createUser(signUpDto);
+    }
+
+    @PostMapping("/login")
+    public SignInUserResponseDto loginUser(@RequestBody SignInUserDto user) throws AuthenticationException {
+        User myUser = userService.findUserByUsername(user.getUsername());
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Business logic should be moved to service.

File: src/main/java/com/sigma/controller/UserController.java [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (passwordEncoder.matches(user.getPassword(), myUser.getPassword())) {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Let's check fail-fast here.

File: src/main/java/com/sigma/controller/UserController.java [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+        throw new AuthenticationException(FAILURE);
+    }
+
+    @DeleteMapping("/users/{user_id}/delete")
+    public String deleteUser(@PathVariable("user_id") Long userId) {
+        userService.deleteUser(userId);
+        return "deleted user";
+    }
+
+    @GetMapping("/admin")
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Unused code

File: src/main/java/com/sigma/dto/LocationDto.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        return new LocationDto()
+                .setId(location.getId())
+                .setLocationName(location.getLocationName())
+                .setStreet(location.getStreet())
+                .setCity(location.getCity())
+                .setHouseNumber(location.getHouseNumber())
+                .setZipCode(location.getZipCode());
+    }
+
+    public Location toLocation() {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/dto/LocationDto.java]: Should be static as well.

File: src/main/java/com/sigma/model/Location.java [hunk 24]
```diff
@@ -16,6 +18,8 @@
 @Entity
 @Table(name = "locations")
 @Data
+@Builder
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/model/Location.java]: Better use Accessors(chain = true)

File: src/main/java/com/sigma/service/impl/LocationServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public Location createLocation(LocationDto location) {
+        log.info("Creating new location {}", location.toString());
+        return locationRepository.save(location.toLocation());
+    }
+
+    @Override
+    public void updateLocation(LocationDto updatedLocation, Long locationId) {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/service/impl/LocationServiceImpl.java:48]: Consider using transactional

File: src/main/java/com/sigma/controller/UserController.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @PostMapping("/login")
+    public SignInUserResponseDto loginUser(@RequestBody SignInUserDto user) throws AuthenticationException {
+        User myUser = userService.findUserByUsername(user.getUsername());
+
+        if (passwordEncoder.matches(user.getPassword(), myUser.getPassword())) {
+            log.info(SUCCESS);
+
+            String token = JWTUtil.generateJWT(myUser, secret, timestamp, issuer);
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/controller/UserController.java]: Timestamp in minutes. TimeUnit.MINUTES.toMillis(valueInMin)

File: src/main/java/com/sigma/filter/AuthFilter.java [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+//    private final AuthenticationManager authenticationManager;
+//
+//    @Override
+//    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) throws IOException, ServletException {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/filter/AuthFilter.java]: `@Component
@Slf4j
@RequiredArgsConstructor
public class AuthFilter extends OncePerRequestFilter {

    @Bean
    public FilterRegistrationBean<AuthFilter> someFilterRegistration() {
        final FilterRegistrationBean<AuthFilter> registration = new FilterRegistrationBean<>();
        registration.setFilter(this);
        registration.addUrlPatterns("!/login/*", "!/register/*");
        registration.setName("authFilter");
        registration.setOrder(Integer.MAX_VALUE);
        return registration;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        final String jwt = ofNullable(request.getHeader("Authourization")).orElse("").replaceAll("Bearer ", "");
        JWTUtil.verify(jwt, "", "");
        filterChain.doFilter(request, response);
    }
}`

File: src/main/java/com/sigma/Application.java [hunk 28]
```diff
@@ -2,10 +2,29 @@
...
     }
-}
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder(BCryptPasswordEncoder.BCryptVersion.$2B);
+    }
+//
+//    @Bean
+//    public FilterRegistrationBean<AuthFilter> someFilter() {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/Application.java]: Remove commented code

File: src/main/java/com/sigma/configuration/AuthFilter.java [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    protected boolean shouldNotFilter(HttpServletRequest request) {
+        return (request.getRequestURI()).matches("(/login)|(/signUp)");
+    }
+
+    @Bean
+    public FilterRegistrationBean<AuthFilter> someFilter() {
+        final FilterRegistrationBean<AuthFilter> registrationBean = new FilterRegistrationBean<>();
+        registrationBean.setFilter(this);
+        registrationBean.addUrlPatterns("/*");//("/(?!.*login).+");
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/AuthFilter.java]: Remove comments, also you mentioned moving this config to Applicaiton class

File: src/main/java/com/sigma/configuration/Config.java [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+//import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+//import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+//import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+//
+//@Configuration
+//@EnableWebSecurity
+//public class Config extends WebSecurityConfigurerAdapter {
+//
+//    @Override
+//    protected void configure(HttpSecurity http) throws Exception {
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/Config.java]: File should be deleted

File: src/main/java/com/sigma/configuration/AuthFilter.java [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        final String jwt = ofNullable(request.getHeader("Authorization")).orElseThrow(() -> new NoSuchElementException());
+        JWTUtil.verify(jwt, secret, issuer);
+        filterChain.doFilter(request, response);
+    }
+
+    @Override
+    protected boolean shouldNotFilter(HttpServletRequest request) {
+        return (request.getRequestURI()).matches("(/login)|(/signUp)");
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/AuthFilter.java]: Should be constants. Better keep public URLs as list to be comfortable in extending it

File: src/main/java/com/sigma/configuration/AuthFilter.java [hunk 32]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Value("${jwt-settings.secret-key}")
+    private String secret;
+
+    @Value("${jwt-settings.issuer}")
+    private String issuer;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        final String jwt = ofNullable(request.getHeader("Authorization")).orElseThrow(() -> new NoSuchElementException());
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/AuthFilter.java]: Use constants

File: src/main/java/com/sigma/JWTUtil.java [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+        JWTVerifier verifier = JWT.require(algorithm)
+                .withIssuer(issuer)
+                .build();
+
+
+        verifier.verify(token.substring(7));
+
+        DecodedJWT jwt = JWT.decode(token.substring(7));
+        if( jwt.getExpiresAt().before(new Date())) {
+            log.error("Token expired");
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/JWTUtil.java:43]: Logging error is not enough. Exception should be thrown

File: src/main/java/com/sigma/configuration/AuthFilter.java [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+        JWTUtil.verify(jwt, secret, issuer);
+        filterChain.doFilter(request, response);
+    }
+
+    @Override
+    protected boolean shouldNotFilter(HttpServletRequest request) {
+        String  regex = "("+constUrls[0]+")";
+        for (int i = 1; i < constUrls.length;i++) {
+            regex+="|("+constUrls[i]+")";
+        }
```
Comment (tarasVarvariuk) [inline src/main/java/com/sigma/configuration/AuthFilter.java]: Not the most elegant way. First of all you can store them in list
`Arrays.asList("login", "signUp")`
Second 'constUrls' is not the best name, probably publicUrls will be better.
Lastly you can join them in much more elegant way:
`publicUrls.stream().map(pu -> String.format("(%s)", pu)).collect(Collectors.joining"|")`
